stan/math/rev/arr/functor/coupled_ode_system.hpp

```diff
@@ -130,6 +130,8 @@ struct coupled_ode_system<F, double, var> {
         }
 
         set_zero_all_adjoints_nested();
+        for (size_t j = 0; j < M_; ++j)
```

Comment:
Can you write out a comment that explains why this happens? It doesn't have to be here; it might be in describing the autodiff variable.

Something to consider... what if we double-nested?

Also, just cause I'm not certain, is this safe? If an exception is thrown while this is nested, what happens? Will a recover memory actually do the right thing if this is left in a nested state?

Reply:
I think it is safe, yes.

If an exception is thrown during the nested AD sweep then we catch that, recover the nested memory and re-throw. The `theta_copy_` is placed on the outer AD stack and will get cleaner by the outer recovery operation (whatever that is).

Reply:
This still needs a comment explaining why this for loop happens, see initial comment.

Reply:
I just realized that we might be able to do this a little cleaner, in a more-understandable way. Would it be possible to do this by `start_nested()` once, putting the `theta_no_chain_` on there, then `start_nested()` repeatedly like we currently do?

Reply:
This would work, yes. But then we would instantiate a coupled ode system object and change the ad stack. That’s rather unclean to me, but this was an earlier implementation I did which I dismissed. Instantiating an object and messing with the ad stack seems not good to me.

Reply:
I'm not sure I understand what you're saying (or I'm missing the nuance).

If I think I understand what you're saying, then this shows that it's not what you think: (start nested) https://github.com/stan-dev/math/blob/502487c511594ccb93eb979df5b8fe163becb417/stan/math/rev/core/start_nested.hpp

I'm really saying that we should not leave the `theta_copy_` on the stack anyway. We should clean up after ourselves. So... on construction, we'd `start_nested()`, then add the `theta_copy_` to the nested stack, do the same thing for each Jacobian call by nesting and unnesting. And then... when we construct the final vars to return, we should `stop_nested()` first, removing `theta_copy_` from the stack. It sounds like that's not what happens and instead, we do everything the same, but leave `theta_copy_` on the stack. I don't think we need to mess with the ad stack any more or less than what we already do.

Reply:
I am afraid that we cannot clean up `theta_nochain_` ... and I think that is ok to me at least (we are using the AD stack here, so there can be side-effects).

If you make a `start_nested` call in the constructor of `coupled_ode_system` then you can get quite weird behaviour if you use `coupled_ode_system` on it's own:

```cpp

var v1 = 2;
var v2 = 4;
 {
   // construction of coupled ode will globally modify the AD stack
    coupled_ode_system mysystem(...);
   // how will this now work??
   v2 *= v1;
   // destruction of the coupled_ode_system will unnest the stack globally
}
// ... I am not sure what happens for v1 & v2 ...
```

So constructing a `coupled_ode_system` globally modifies the AD stack...I don't think this is a good thing to me. Leaving a few bits behind on the AD stack is the better choice, I think.

If you want to avoid leaving things behind, then we can put inside the `integrate*` function the `coupled_ode_system` into a nested block. That can be cleaned up.

Reply:
This last comment you made really scares me. It really makes me think that it's just setting us up for a disaster. I understand that we modify the global stack, but to have a side-effect where things are left in an improper state is a bit too risky for developer error.

> If you want to avoid leaving things behind, then we can put inside the integrate* function the coupled_ode_system into a nested block. That can be cleaned up.

This makes it sound like you know how to remove that behavior? Let's do it if possible! If it's not, let's reevaluate it. (To me, this is new information that this change really leaves the autodiff stack in an unsafe state between different operations and is the first one I've heard of in the Math library.)